Feature discoverable by pattern.

miguelius opened the pull request:
This is WIP. Not to merge yet. I couldn't run the tests because I am working on Linux and they don't run with dotnet core 6 runtime.

It pretends to solve issue #64.

There are some issues regarding to the way that Feature Class is coupled to Feature File. To bypass this, we send it as an argument when we discover the features, but i would be nicer to have a better solution.

If you build the dll and add it to an existing project, in the FeatureClass you can add something like this:

```c#
[FeatureFileSearchPattern("./Features/ConsultaHuerfanos*.feature")]
public sealed class ConsultaHuerfanosFeature : Feature
```
And it should work find the files and bind it to the Feature.

By default it searches for ClassName*.feature.

Improvements, reviews and ideas with the fork are welcome.

Cheers!

ttutisani commented:
What will happen if two files match the pattern?

miguelius commented:
It's exactly the scenario we are looking for. We'd like to have many feature files sharing the steps of a Feature Class.

It keeps them tracked with the extra-argument in the execution.

I am adjusting the texts to the discovery model.

miguelius commented:
Tests fixed.

Main changes: Discover now returns Featue instance and file. Since executions parses them from the filles we had to passed its path.

We return a tuple, but it would be nice to have some derived class from feature including the path to trace it or some other structure.

In the current state you can:

- define a class in the classic way
- define using the searchpath attribute
- use both

The model will discover all the files matching and return in a similar way of the missing features.

We are giving this a test drive with a local project and refactors are welcome. We will provide the tests for the described behaviour.

I understand it can be improved but we wanted to give back to your community and you our humble work. This is a way of saying thank you.

It's a great project and very well documented and maintained! Congrats

Cheers!

miguelius commented:
Of course!

Say you have the the ComplexFeature where many business experts are working, with many aspects. While the feature has a set of steps, different combinations can be made.

You've written a small runtime where you just write the feature file in the folder Features. The runtime starts scans the features assembly and you have Complex.feature file in the Features folder. In order to simplify the existence of the business analysts, it would be nice to have a couple of feature files associated to the same CompleFeature class. For instance ComplexGroupOfScenarios1.feature, ComplexGroupOfScenarios2.feature and ComplexGroupOfScenarios3.feature would be files that describe scenarios for this common Feature. Each analyst work in different files, preventing conflicts between them.

Let's go to code:

```c#
 [FeatureFileSearchPattern("Features/*Complex*.pattern")]
class ComplexFeature : Feature {

  // series of given, when, then shared ammong scenarios and shared state within a concret feature

}
```

In the directory Features we have:

ComplexGroupOfScenarios1.feature file:
```gherkin
Feature Complex scenarios concerning some aspect
...
  Scenario 1
...
  Scenario 2
...
```
ComplexGroupOfScenarios2.feature file:
```gherkin
Feature Complex scenarios concerning some other aspect
...
  Scenario 1
... 
  Scenario 2
...
```

ComplexGroupOfScenarios3.feature file:
```gherkin
Feature Complex scenarios concerning some other creepy and super important aspect
...
  Scenario 1
... 
  Scenario 2
...
```

Analysts can write as many Complex*.pattern as they need without having conflicts.

I understand that Features shouldn't be thaaat big to have as many different scenarios. And you could also generate ComplexGroupOfScenarios1Feature, ComplexGroupOfScenarios2Feature and ComplexGroupOfScenarios3Feature classes extending Feature and the effect would be the same... Once we have compiled and redistritbuted the Runtime. This way with one common Feature class we can have many Files sharing the scenarios with the same Runtime.

The default pattern is ClassName*.feature. And it keeps the compatibility with FeatureFile Attribute. It also prevents loading it twice if it matches the pattern.

I'll write a test case describing a similar situation.

miguelius commented:
Hi! We've added a couple of test cases.

1. AddFeature with two feature files: AddTwoNumbers.feature, AddNumbersTo5.feature
2. ComplexFeature with 4 feature files: Features/ComplexGroupOfScenarios1.feature, Features/ComplexGroupOfScenarios2.feature, Features/ComplexGroupOfScenarios3.feature, Features/Complex.feature.

Each feature file has different scenarios with an scenario name identifying the originating file.

Cheers!

jclcbrain commented:
Hi,

I want to support this feature request. 

We are using XUnit.Gherkin.Quick to create a small test-suite for domain-specific software. We want our non-coders (but still quite technical personel) to be able to create new .feature files without creating a new feature class, but utilizing an already existing feature class with generic step definitions.

The feature class is then built and shipped and our test-writers are able to create feature files to use with the feature class.

As of now, the need to create a feature class and build the solution does not support this use case. 
This pull request seems create functionality to support our use case though.
We are very excited about this.

ttutisani commented:
Thank you all for your input!

@miguelius Did you push your commit? I don't see any such test cases when I open the files changed in this PR. Am I missing something?

I assume that you added those test cases under the Xunit.Gherkin.Quick.ProjectConsumer folder, but I don't see any changes under that folder. I only see changes to the unit tests and the Xunit.Gherkin.Quick itself.

![image](https://user-images.githubusercontent.com/18409870/190500319-218a670e-4642-428b-bce1-691765a1d1aa.png)

miguelius commented:
I've add an extra file to Xunit.Gherkin.Quick.ProjectConsumer/Addition_ForMultipleUseCases/AddTwoNumbersExtra.feature and the pattern search to Addition_ForMultipleUseCases/AddTwoNumbers.cs.

I've fixed an issue. When I added a class and a MissingFeature class was present, it assumed feature files associated by pattern were not implemented. So, it run the scenarios and also printed they were not implemented.

You can give it a try and if there is another issue, we'll gladly fix it.

ttutisani commented:
I downloaded your code and ran the tests. I saw a strange thing - the tests under ProjectConsumer.Addition.AddTwoNumbers show duplicate results in the test explorer. That means the same test is executed multiple times. The only thing that is not duplicated is the new test that you added (AddTwoNumbersExtra). See the picture below. Please fix this to ensure the execution is fine and then I can review it.

![image](https://user-images.githubusercontent.com/18409870/190781915-3563ee5b-4aa6-46c0-b8e7-3d9d883ca570.png)

If you try to pull the version before your changes, the same tests are not duplicated, so it looks like the duplication appears after your commits.

miguelius commented:
I think I've fixed the bug. Or at least I see fewer test cases. The issues was when it was preventing to include the pattern and the explicit file name it compared by the string literal. Now, it uses the absolute path to compare filenames.

Please review it, and of course any other thing you may find, we'll be happy to fix it or improve it. Thanks for your review!

ttutisani commented:
Thanks for fixing that!
I see another strange behavior, is it me or for your is the same? In this PR, I can't find the "FeatureFileSearchPatternAttribute.cs" file. It is there in your fork but I can't find it in the PR, so I can't review it. Maybe you need to close and reopen the PR, I don't know. I hope you can figure it out and let me know when all files are in the PR.

ttutisani commented:
Ah, never mind, I didn't realize this class existed before. Completely forgot. I will review the rest. Thanks!

ttutisani commented on source/Xunit.Gherkin.Quick/CoreModel/Feature/Discovery/FeatureDiscoveryModel.cs:
```diff
@@ -11,18 +13,31 @@ public FeatureDiscoveryModel(IFeatureFileRepository featureFileRepository)
-            if (featureFile == null)
-                throw new System.IO.FileNotFoundException("Feature file not found.", fileName);
+            var featureClassInfo = FeatureClassInfo.FromFeatureClassType(featureClassType);
+            var fileNameSearchPattern = featureClassInfo.FileNameSearchPattern;
```
I don't follow. You are not using the fileNameSearchPattern in this method, but you should.

What you should have in this method is this:
1. Find the feature file using featureClassInfo.FeatureFilePath (this already worked, so it's already there before your changes).
2. Find feature files using featureClassInfo.FileNameSearchPattern.
3. Union 1 and 2 from above.
4. Create Gherkin.Ast.Feature for all those files. and return that.

What you have works perhaps because you passed the file name search pattern into the repository, which is a bad idea because you have created tight coupling between the discoverers and this class. The caller should not help the called method do its job, that's tight coupling. All the logic must be here, not outside.

ttutisani commented on source/Xunit.Gherkin.Quick.ProjectConsumer/Addition_ForMultipleUseCases/AddTwoNumbers.cs:
```diff
@@ -1,6 +1,7 @@
 {
-    [FeatureFile("./Addition_ForMultipleUseCases/AddTwoNumbers.feature")]
+    [FeatureFile("Addition_ForMultipleUseCases/AddTwoNumbers.feature")]
+    [FeatureFileSearchPattern("Addition_ForMultipleUseCases/AddTwoNumbers*.feature")]
```
Also, please show in this class methods how the filename is received and where. I don't see its usage so I cannot say if it looks good.

miguelius replied:
First of all, thank you for your review!

I'll try to use the FeatureFile instead of using the FeatureFileSearchPattern.

ttutisani replied:
Whichever you prefer will be fine, either a new attribute class or the same one. In fact, if you look at the feature file path, it is also a pattern, without an asterisk and an exact match hence. That is why I suggested.

ttutisani commented on source/Xunit.Gherkin.Quick/CoreModel/Feature/FeatureClass.cs:
```diff
@@ -9,6 +9,7 @@ namespace Xunit.Gherkin.Quick
     internal sealed class FeatureClass
     {
         public string FeatureFilePath { get; }
+        public string FileNameSearchPattern { get; }
```
Please rename to "FeatureFileSearchPattern" for clarity, otherwise it's confusing what file it is talking about.

ttutisani commented on source/Xunit.Gherkin.Quick/CoreModel/Feature/FeatureClassInfo.cs:
```diff
@@ -7,8 +7,14 @@ internal sealed class FeatureClassInfo
         public string FeatureFilePath { get; }
 
-        private FeatureClassInfo(string featureFilePath)
+        public string FileNameSearchPattern { get; }
```
Same here. Please rename to "FeatureFileSearchPattern"

ttutisani commented on source/Xunit.Gherkin.Quick/CoreModel/Feature/FeatureClassInfo.cs:
```diff
@@ -7,8 +7,14 @@ internal sealed class FeatureClassInfo
         {
+            FileNameSearchPattern = !string.IsNullOrWhiteSpace(fileNameSearchPattern)
+                ? fileNameSearchPattern 
+                : throw new System.ArgumentNullException(nameof(fileNameSearchPattern));
```
Move this one blow down, to be in the same order as the arguments received.

ttutisani commented on source/Xunit.Gherkin.Quick/CoreModel/Feature/Discovery/MissingFeatureDiscoveryModel.cs:
```diff
@@ -23,10 +23,16 @@ internal sealed class MissingFeatureDiscoveryModel
-            var newFiles = allFiles.Where(f => !referencedFiles.Contains(f))
+                
+            var referencedFilesByPattern = _featureClassInfoRepository.GetFeatureClassesInfo()
+                .SelectMany(fc => new FeatureFileRepository(fc.FileNameSearchPattern).GetFeatureFilePaths())
```
Don't use repository like this, it will violate the single responsibility principle for the repository. Repository is only responsible to find all feature paths and get the feature by path. To find the paths that match the pattern, you need something else. You can create a helper class for that if you need to or write the code inside this method itself. But don't use repository.

ttutisani commented on source/Xunit.Gherkin.Quick/ScenarioXunitHook/Scenario/ScenarioDiscoverer.cs:
```diff
@@ -19,21 +19,26 @@ public ScenarioDiscoverer(IMessageSink messageSink)
+            var featureClassInfo = FeatureClassInfo.FromFeatureClassType(featureClass);
 
-            foreach (var scenario in feature.Children.OfType<global::Gherkin.Ast.Scenario>())
+            var features = new FeatureDiscoveryModel(new FeatureFileRepository(featureClassInfo.FileNameSearchPattern)).Discover(featureClass);
```
This is what I mentioned earlier. This line creates tight coupling between this method and the FeatureDiscoveryModel. This method itself is not responsible to know about the FeatureClassInfo anything. Do that logic inside if necessary but not outside.

ttutisani commented on source/Xunit.Gherkin.Quick/ScenarioXunitHook/Scenario/ScenarioDiscoverer.cs:
```diff
@@ -19,21 +19,26 @@ public ScenarioDiscoverer(IMessageSink messageSink)
+                    yield return new ScenarioXunitTestCase(
+                        _messageSink, 
+                        testMethod, 
+                        feature.Item2.Name, 
```
What is Item2 and Item1? Confusing names.

miguelius replied:
Oh this is an issue of .net tuples. I think it would be better to have an adhoc type instead of using a tuple.

ttutisani commented on source/Xunit.Gherkin.Quick/ScenarioXunitHook/ScenarioOutline/ScenarioOutlineDiscoverer.cs:
```diff
@@ -20,30 +20,36 @@ public ScenarioOutlineDiscoverer(IMessageSink messageSink)
+            var featureClassInfo = FeatureClassInfo.FromFeatureClassType(featureClass);
 
-            foreach (var scenarioOutline in feature.Children.OfType<ScenarioOutline>())
+            var features = new FeatureDiscoveryModel(new FeatureFileRepository(featureClassInfo.FileNameSearchPattern)).Discover(featureClass);
```
Same here. Tight coupling and also code duplication, which proves the point. If you move the logic inside, you will not need to do the same thing in two places.

ttutisani commented on source/Xunit.Gherkin.Quick/ScenarioXunitHook/ScenarioOutline/ScenarioOutlineDiscoverer.cs:
```diff
@@ -20,30 +20,36 @@ public ScenarioOutlineDiscoverer(IMessageSink messageSink)
+                        foreach (var row in example.TableBody)
+                        {
+                            var tags = feature.Item2.GetExamplesTags(scenarioOutline.Name, example.Name);
+                            var skip = feature.Item2.IsExamplesIgnored(scenarioOutline.Name, example.Name);
```
What is this Item2 and item1? Is it possible to avoid using them?

miguelius replied:
Yes, it's possible to avoid them by an ad hoc type. Same as above, it's because of .net tuples.

ttutisani reviewed (commented):
Thanks for your PR! I was able to understand what you tried to do. Please fix the comments and I will come back to this.

miguelius commented on source/Xunit.Gherkin.Quick/CoreModel/Feature/Discovery/FeatureDiscoveryModel.cs:
```diff
@@ -11,18 +13,31 @@ public FeatureDiscoveryModel(IFeatureFileRepository featureFileRepository)
         }
 
-        public global::Gherkin.Ast.Feature Discover(Type featureClassType)
+        public IEnumerable<Tuple<string,global::Gherkin.Ast.Feature>> Discover(Type featureClassType)
```
I've choosen IEnumerable since many Xunit Library such as the Discoverer which invokes it, uses IEnumerable. 

Since I am not using anything from the List semantic, I return the less restrictive type. Also it prevents anyone from adding elements.

ttutisani replied:
I understand your point, but it violates robustness principle: it's actually better to return concrete type and take general type from the code design perspective. Xunit is not an exemplary design reference in this case, not the best one to follow, with all due respect.
Also, when returning IEnumerable, the list is not materialized inside the method, but instead where the enumeration is necessary. If the materialization causes the exception to be thrown, the stack trace will point at the materialization line in code, not this method, so it's a maintenance headache as well.
Adding the item to the list is not a concern of the method, the caller needs to take care of the returned value, so that is also not a problem.
With all these points, please change it to List.

miguelius commented:
Thank you very much for review! We'll be working on it. Besides the implementation aspects mentioned above, have you found any functional Issue? If so, please, you are welcome to tell us.

ttutisani commented:
Functionality looked good when comparing it before and after the change, so it's all good 👍

miguelius commented:
Hi!

Fixed most of your comments.

Also changes the way it's used:

```csharp
[FeatureFile("Features/*Complex*.feature")]
class ComplexFeature : Feature {

  // series of given, when, then shared ammong scenarios and shared state within a concret feature

}
```

Or in case of AddNumber:

```csharp
[FeatureFile("Features/Add*.feature")]
class AddFeature : Feature {

// same here!

}
```

Classes are more or less the same than the original.

- FeatureClassInfo has a property to know wether you are working with file path or a pattern
- FeatureClassInfo has a matcher to determine if the file you're trying is treated by the feature. This allows MissingFeature discoverer to know if the file is a handled Feature File.
- The discover model was fixed to return a List.

Any other comment or suggestions is welcomed.

Cheers!

ttutisani commented on source/Xunit.Gherkin.Quick/CoreModel/Feature/Discovery/FeatureDiscoveryModel.cs:
```diff
@@ -1,28 +1,60 @@
+                .Select(f => new FeatureFile(f, _featureFileRepository.GetByFilePath(f).GherkinDocument.Feature))
+                .ToList() ?? new List<FeatureFile>();
+
+            if (newFeatures.Count == 0) {
```
Move this inside the If statement above, right after featureFilePaths.AddRange, that will make more sense for the future code readers. Also, no computation above will be done if featureFilePaths.Count==0.

ttutisani commented on source/Xunit.Gherkin.Quick/CoreModel/Feature/FeatureClassInfo.cs:
```diff
@@ -1,17 +1,28 @@
     {
         public string FeatureFilePath { get; }
+        private Regex _matcher;
+        public bool IsPattern { get { return FeatureFilePath.Contains("*"); } }
```
This getter is too expensive - string operation is more expensive than null check. So here, `return _matcher !== null` is better.

miguelius replied:
Done!

ttutisani commented on source/Xunit.Gherkin.Quick/CoreModel/Feature/FeatureClassInfo.cs:
```diff
@@ -1,17 +1,28 @@
                 ? featureFilePath
                 : throw new ArgumentNullException(nameof(featureFilePath));
+
+            if (this.IsPattern) {
```
Since you will change the above IsPattern getter, here, you can check `featureFilePath.Contains("*")`, which is better here because it will only be done once.

miguelius replied:
Good idea! Thanks!

ttutisani reviewed (commented):
You've done a tremendous job here! Thanks for the improvements!
We are almost there. Please address these small comments and we will be good to merge.
👍

miguelius commented:
Thanks for the feed back, Tengiz!

I've added both corrections! To prevent having two return points I used a dictionary in Discover method. Hope it's fine. If not, improvements are welcome!

Cheers!
